iob_soc.py

```diff
@@ -93,49 +93,55 @@ def _create_instances(cls):
-                {"interface": "clk_en_rst_s_port"},
-                iob_merge,
-                iob_split,
-                iob_rom_sp,
-                iob_ram_dp_be,
-                iob_ram_dp_be_xil,
-                iob_pulse_gen,
-                iob_counter,
-                iob_reg,
-                iob_reg_re,
-                iob_ram_sp_be,
-                iob_ram_dp,
-                iob_reset_sync,
-                iob_ctls,
-                axi_interconnect,
-                # Simulation headers & modules
-                (axi_ram, {"purpose": "simulation"}),
-                ({"interface": "axi_s_portmap"}, {"purpose": "simulation"}),
-                (iob_tasks, {"purpose": "simulation"}),
-                # Software modules
-                printf,
-                # Modules required for CACHE
-                (iob_ram_2p, {"purpose": "simulation"}),
-                (iob_ram_2p, {"purpose": "fpga"}),
-                (iob_ram_sp, {"purpose": "simulation"}),
-                (iob_ram_sp, {"purpose": "fpga"}),
-            ]
-            + extra_submodules
-        )
+        submodule_list = [
+            iob_picorv32,
+            iob_uart,
+            iob_timer,
+            # Hardware headers & modules
+            {"interface": "iob_wire"},
+            {"interface": "axi_wire"},
+            {"interface": "axi_m_port"},
+            {"interface": "axi_m_m_portmap"},
+            {"interface": "axi_m_portmap"},
+            iob_utils,
+            {"interface": "clk_en_rst_s_s_portmap"},
+            {"interface": "clk_en_rst_s_port"},
+            iob_merge,
+            iob_split,
+            (iob_rom_sp, {"purpose": "simulation"}),
+            (iob_rom_sp, {"purpose": "fpga"}),
+            (iob_ram_dp_be, {"purpose": "simulation"}),
+            (iob_ram_dp_be, {"purpose": "fpga"}),
+            (iob_ram_dp_be_xil, {"purpose": "simulation"}),
+            (iob_ram_dp_be_xil, {"purpose": "fpga"}),
+            iob_pulse_gen,
+            iob_counter,
+            iob_reg,
+            iob_reg_re,
+            (iob_ram_sp_be, {"purpose": "simulation"}),
+            (iob_ram_sp_be, {"purpose": "fpga"}),
+            (iob_ram_dp, {"purpose": "simulation"}),
+            (iob_ram_dp, {"purpose": "fpga"}),
+            iob_reset_sync,
+            iob_ctls,
+            axi_interconnect,
+            # Simulation headers & modules
+            (axi_ram, {"purpose": "simulation"}),
+            ({"interface": "axi_s_portmap"}, {"purpose": "simulation"}),
+            (iob_tasks, {"purpose": "simulation"}),
+            # Software modules
+            printf,
+            iob_cache,
+            # Modules required for CACHE
+            (iob_ram_2p, {"purpose": "simulation"}),
+            (iob_ram_2p, {"purpose": "fpga"}),
+            (iob_ram_sp, {"purpose": "simulation"}),
+            (iob_ram_sp, {"purpose": "fpga"}),
+        ]
+        if "USE_EXTMEM" not in sys.argv:
+            submodule_list.remove(iob_cache)
```

Comment:
instead of adding cache and then removing,
only add cache and cache memory submodujles if necessary:
```python
if "USE_EXTMEM" in sys.argv:
    submodule_list += [
        iob_cache,
        # Modules required for CACHE
        (iob_ram_2p, {"purpose": "simulation"}),
        (iob_ram_2p, {"purpose": "fpga"}),
        (iob_ram_sp, {"purpose": "simulation"}),
        (iob_ram_sp, {"purpose": "fpga"}),
    ]
```

Reply:
change made